### ISSUE TITLE
S3Queue table is producing listing calls excessively

### ISSUE THREAD

**Issue by hoa-grvt** [ClickHouse/ClickHouse#65477]
**Describe the unexpected behaviour**

I have a S3Queue table like so: 

```
CREATE TABLE msg_type_queue (
   [Redacted]
)
ENGINE = S3Queue('{{.Endpoint}}/{{.Bucket}}/type=msg_type/year=*/month=*/day=*/hour=*/minute=*/*.bin', 'CapnProto')
SETTINGS
  mode = 'ordered'
```

We have about **7** S3Queue total with different `type=msg_type_{1..7}`. And about **30** downstream materialized viewer (no heavy processing, just select different fields from the queues created above). We generate maximum 1 file per second for each msg type, so there are maximum of `60*60*24` files per day per msg type per path. We observed about 6,000 listing calls per minute - stats: 99% listing call API and 1% get call API on our S3 bucket

**How to reproduce**
* Which ClickHouse server version to use: `v24.5.1.1763-stable`
* Create a table with an S3Queue engine
* Generate a day worth of files following the path filter pattern above
* See the listing calls statistics and bill from AWS console the next day.

**Expected behavior**
ClickHouse should be polling at a more reasonable rate or rather path prefix filtering in listing should be more performant.

On our part, we are not sure if our prefix pattern is suitable for clickhouse s3queue workload as well.

**Additional context**
N/A

**Comment by kssenii** [comment]
It should do not more than 1 list request per second by default. The bahaviour is controllable by settings `s3queue_polling_min_timeout_ms (== 1000 by default)`, `s3queue_polling_max_timeout_ms (== 10000 by default)`. You can try to increase them.

**Comment by hoa-grvt** [comment]
@kssenii Thank you for your response. We've double-checked our config via show create table. And it's using the default config (1s polling min timeout and 10s polling max timeout).

Hmm, is this because of s3 list api  is limited to 1000 items per call, and when the path filter has many items it will result to multiple listing requests per call? I suppose there is no prefix + delimiter option to set (the number of listing call will roll-up to 1 if those options are set) as per (docs)[https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-prefixes.html].

Increase timeout for polling doesn't seem like an optimal option / solution for us (more latency as the result)

**Comment by ckevin** [comment]
It would be very useful to be able to populate the S3Queue via bucket notifications from SQS, instead of from polling.  Our ideal scenario would be one SQS queue demuxing into all our different S3Queue tables, so that each new bucket would just require setting up notifications to the existing queue, along with the S3Queue structure on the CH side.

If we can't do that, another useful option would be to limit the scan range by templating the S3 prefix string with escape sequences for current year/month/day.  There would be edge cases around day rollovers, but that could be resolved by scanning more than one instantiation of the prefix (i.e. for both current & prior day) for some amount of overlap time.

**Comment by kssenii** [comment]
@hoa-grvt 

> is this because of s3 list api is limited to 1000 items per call, and when the path filter has many items it will result to multiple listing requests per call?

Yes, this is true, but the settings I mentioned are not affecting these list requests. They affect only separate iterations of list requests, which might consist of many list requests split by 1000 items. We have a setting to define this batch size - `s3_list_object_keys_size` (which is by default 1000). May be this setting if what you need. We did not add timeout in between list batch requests, because it would affect s3queue load speed.

> Increase timeout for polling doesn't seem like an optimal option / solution for us (more latency as the result)

Could you suggest something else?

BTW please do not hesitate to ping again if there is no answer on the issue which was already answered before. I sometimes miss notifications from github and do not see new replies.

**Comment by kssenii** [comment]
@ckevin 

> It would be very useful to be able to populate the S3Queue via bucket notifications from SQS, instead of from polling. Our ideal scenario would be one SQS queue demuxing into all our different S3Queue tables, so that each new bucket would just require setting up notifications to the existing queue, along with the S3Queue structure on the CH side.

This seems to be a big work to investigate and to do, while this is a quite limited scenario which is not for everybody. So I doubt we would implement it ourselves, but a pull request is welcome if you want to implement it yourself.

**Comment by ckevin** [comment]
> @ckevin
> 
> > It would be very useful to be able to populate the S3Queue via bucket notifications from SQS, instead of from polling. Our ideal scenario would be one SQS queue demuxing into all our different S3Queue tables, so that each new bucket would just require setting up notifications to the existing queue, along with the S3Queue structure on the CH side.
> 
> This seems to be a big work to investigate and to do, while this is a quite limited scenario which is not for everybody. So I doubt we would implement it ourselves, but a pull request is welcome if you want to implement it yourself.

Thanks for responding!  We might get to that point eventually.  Right now we're doing some evaluation of locally hosted CH, and this is one of the things that came up with S3 data onboarding.  We frequently want to auto-ingest new files from prefixes that already have hundreds of thousands, or millions, of files in them.  It's fairly simple on our side to setup SQS queue(s) and S3 "new object" notifications, which create queues of just the new filenames -- if we could use that information, the same historic files wouldn't need to be enumerated over and over again.  What we don't have is a way to give that information to CH.  The simplest implementation for us might be to setup a Lambda which reads those notifications (either triggered or periodically in small batches), and loads the files to CH as they appear (instead of using S3Queue.)  But that's basically what S3Queue is already doing in Keeper, just via file enumeration instead of from a notification queue.

CH implementation might not be as bad as you think.  Since S3Queue is already doing periodic polling in Keeper, the difference might just be reading an SQS queue for file create notifications, instead of the S3 API to enumerate all matching filenames.  I am not well versed in the S3Queue implementation code (very new to CH here), but I would guess that the simplest thing to support would be a separate SQS ARN per S3Queue table, which would contain notifications only the files scoped for that table.  (It would be nice to have demuxing from one SQS queue to multiple S3Queue, but we'd gladly implement a separate queue per S3Queue table to get orders of magnitude better efficiency on new file discovery.)

The bit I mentioned before about supporting year/month/date templating for S3Queue prefixes would also be helpful, if that's easier.  We can control what our S3 prefixes look like, so if our files go into buckets with `/%Y/%m/%d/` and S3Queue could be configured via templating to read only the prefix that matches today's date (probably with some small overlap to make sure no files are missed around day boundaries) that would make the polling a lot more efficient.

**Comment by hoa-grvt** [comment]
@kssenii thank you for your reply.

I think the main performance driver should be done on the CH side to allow user to specify `Delimiter` and `Prefix` shown in here:
https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-prefixes.html . 

It will roll-up all call/request with specified prefix/delimiter as 1 single call, that's a huge performance win in term of cost for non-flat key namespace.

**Comment by azat** [comment]
As far as I can see (@kssenii, please correct me if I am wrong), this should only happen when everything has been processed. Otherwise, S3Queue should issue requests only after all files in a given request have been processed and should continue iterating without restarting from scratch until everything is processed or an error occurs.

Additionally, this should not be a problem at all if you are using `after_processing=delete`.
